### PR TITLE
feature/174 - regex cleanup

### LINKED ISSUE #174: Improve `toSafeRegex` Performance
`toSafeRegex` has some issues that lead to slow queries

 * uses `.*`instead of `.*?` which causes recursive backtracking - it's a known regex engine performance killer
   * lots of articles on the web about this -- here's a recent one: https://javascript.info/regexp-catastrophic-backtracking
 * I see no need for the match on each letter - the elastic regex engine must support case insensitive matches. Will investigate why it was ever done this way, though

### PR DESCRIPTION
**Pull request by doug-patterson**
Fixes #174 

### Description
* Update from out hand-rolled way of doing a case-insensitive regex match to the native ES7 supported version.

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>The version was not updated. Please update the version.</td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>The changelog has not been updated. Please update the changelog.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The README has not been updated. Please update the README.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1a05dcac0b2f255fcba89b8939f8765457ff09bb
</p>

**Comment by doug-patterson** [comment]
OK, I like the changes but there are 14 unit tests to fix, WIP on that.

Note: these changes can't break anything because (1) they only affect what substrings will be declared to match a given string and not whether a match is found and (2) we only use these function to filter values, for which it only matters which strings contain a match for what the user types and not what it is.

Note on Note: the substring match will only actually change here in vanishingly few cases anyway -- the best real world scenarios I can think of involve "New York New York" etc. Again, all that matters to the options filter, etc is whether there is a match.

The performance improvement here is that if you type "new york" into the facet options filter it'll be satisfied by the first two words and stop looking -- for every one of the potentially millions of records it has to check. On the existing way of doing it when you type that for each record it'll find "new york" and then sill need to check whether it actually found the *last* occurrence of "york" in the string, which it'll do by backtracking in V8 and most other engines.

File: src/utils/regex.js [hunk 1]
```diff
@@ -1,22 +1,18 @@
 let _ = require('lodash/fp')
 
-let toSafeRegex = _.flow(
-  _.replace(/[.?+*|{}[\]()]/g, '\\$&'),
```
Comment (doug-patterson) [inline src/utils/regex.js:4]: @daedalus28 I'm not sure what this was for -- will restore if it's needed -- as you can see I've just turned the junk into the null string instead

**Comment by doug-patterson** [comment]
need to check on some facts about the Lucene regex engine actually....

**Comment by doug-patterson** [comment]
yeah, needs an update for Lucene, little but...

File: src/utils/regex.js [hunk 2]
```diff
@@ -1,22 +1,18 @@
...
-  _.map(x => (x.match(/[A-Za-z]/) ? `[${_.toUpper(x)}${_.toLower(x)}]` : x)),
-  _.join('')
-)
+let toSafeRegex = _.flow(_.replace(/[.?+*|{}[\]()]/g, ''))
```
Comment (daedalus28) [inline src/utils/regex.js:3]: We should replace these with spaces to more closely match how es will analyze these. If you search for `one/two`, you’d want it to run `one two` and not `onetwo`

File: src/utils/regex.js [hunk 3]
```diff
@@ -1,22 +1,18 @@
...
 
-let regexPartsForWords = _.flow(
-  _.replace(/\s\s+/g, ' '),
```
Comment (daedalus28) [inline src/utils/regex.js:10]: We should probably still strip extra spaces

File: test/example-types/filters/text.js [hunk 4]
```diff
@@ -150,12 +149,12 @@ describe('text', () => {
...
             {
               regexp: {
-                'description.untouched': '[Ll][Aa][Ss][Ee][Rr][Jj][Ee][Tt]',
+                'description.untouched': 'laserjet',
               },
             },
             {
               regexp: {
-                'description.untouched': '[Pp][Rr][Ii][Nn][Tt][Ee][Rr]',
+                'description.untouched': 'printer',
```
Comment (daedalus28) [inline test/example-types/filters/text.js:157]: These should have the case insensitive flag

**Review by daedalus28 — CHANGES_REQUESTED** [review]

File: src/utils/regex.js (hunk 3 above)
Comment (doug-patterson) [reply src/utils/regex.js:10]: ⛳ : the split on `/s+/` has the same effect

File: test/example-types/filters/text.js (hunk 4 above)
Comment (doug-patterson) [reply test/example-types/filters/text.js:157]: the tests pass and will fail if I add that. Not sure what you could mean here

Comment (daedalus28) [reply test/example-types/filters/text.js:157]: Based on this test, it looks like you're generating a query that isn't using the case insensitive flag but no longer manually does case-insensitivity with the letter brackets. If we leave this test as is, it shows that the behavior will change.

**Review by daedalus28 — APPROVED** [review]

**Comment by daedalus28** [comment]
Just needs release prep - package, changelog, readme

**Pull request merged**